src/compute-types/src/plan/flat_plan.rs

```diff
@@ -657,6 +657,59 @@ impl<T> FlatPlan<T> {
     pub fn destruct(self) -> (BTreeMap<NodeId, FlatPlanNode<T>>, NodeId) {
         (self.nodes, self.root)
     }
+
+    /// Return the list of contained node IDs in topological order (dependencies before
+    /// dependants).
+    ///
+    /// # Panics
+    ///
+    /// Panics if the plan contains cycles.
+    pub fn order_topological(&self) -> Vec<NodeId> {
```

Comment:
I think this wouldn't be needed if we collected the flat plan using a post-order traversal -- everything should be in the correct order in this case. As discussed separately, I'm fine with working on this in a subsequent PR.

Reply:
Post-order flattening has the issue that you cannot move out of the `Plan` nodes while you are still holding on to their parents that you need to flatten later. We could work around this by cloning everything, but I think it would also work to flatten in pre-order (or the variant of pre-order where you visit children right-to-left) and then reversing the order of flattened nodes in the end.

So both visiting orders should allow us to get to a topological ordering of nodes. The reason I'm doing another sort here is that the `FlatPlan` doesn't remember the visiting order, it just throws everything into a map. We can consider storing the `Vec<NodeId>` that defines the topological order as another field in the `FlatPlan`. That might be less complicated than implementing a topological sort, let me try it out.